rust/rope/src/diff.rs

```diff
@@ -0,0 +1,329 @@
+    (start.min(max_left), end)
+}
+
+/// Finds the longest increasing subset of copyable regions. This is essentially
+/// the longest increasing subsequence problem. This implementation is adapted
+/// from https://codereview.stackexchange.com/questions/187337/longest-increasing-subsequence-algorithm
+fn longest_increasing_region_set(items: &[(usize, usize)]) -> Vec<(usize, usize)> {
+    let mut result = vec![0];
+    let mut prev_chain = vec![0; items.len()];
+
+    for i in 1..items.len() {
+        // If the next item is greater than the last item of the current longest
+        // subsequence, push its index at the end of the result and continue.
+        let last_idx = *result.last().unwrap();
+        if items[last_idx].1 < items[i].1 {
+            prev_chain[i] = last_idx;
+            result.push(i);
+            continue;
+        }
+
+        let next_idx = match result.binary_search_by(
+            |&j| items[j].1.cmp(&items[i].1)) {
+                Ok(_) => continue, // we ignore duplicates
+                Err(idx) => idx,
+        };
+
+        if &items[i].1 < &items[result[next_idx]].1 {
+            if next_idx > 0 {
+                prev_chain[i] = result[next_idx - 1];
+            }
+            result[next_idx] = i;
+        }
+    }
+
+    // walk backwards from the last item in result to build the final sequence
+    let mut u = result.len();
+    let mut v = *result.last().unwrap();
+    while u != 0 {
+        u -= 1;
+        result[u] = v;
+        v = prev_chain[v];
+    }
+    result.iter().map(|i| items[*i]).collect()
+}
+
+#[inline]
+fn non_ws_offset(s: &str) -> usize {
+    s.as_bytes()
+        .iter()
+        .take_while(|b| **b == b' ' || **b == b'\t')
+        .count()
+}
+
+/// Represents copying `len` bytes from base to target.
+#[derive(Debug, Clone, Copy)]
+struct DiffOp {
+    target_idx: usize,
+    base_idx: usize,
+    len: usize,
+}
+
+/// Keeps track of copy ops during diff construction.
+#[derive(Debug, Clone, Default)]
+pub struct DiffBuilder {
+    ops: Vec<DiffOp>,
+}
+
+impl DiffBuilder {
+    fn copy(&mut self, base: usize, target: usize, len: usize) {
+        if let Some(prev) = self.ops.last_mut() {
+            let prev_end = prev.target_idx + prev.len;
+            let base_end = prev.base_idx + prev.len;
+            assert!(prev_end <= target, "{} <= {} prev {:?}", prev_end, target, prev);
+            if prev_end == target && base_end == base {
+                prev.len += len;
+                return;
+            }
+        }
+        self.ops.push(DiffOp { target_idx: target, base_idx: base, len: len })
+    }
+
+    fn to_delta(self, base: &Rope, target: &Rope) -> RopeDelta {
+        let mut els = Vec::with_capacity(self.ops.len() * 2);
+        let mut targ_pos = 0;
+        for DiffOp { base_idx, target_idx, len } in self.ops {
+            if target_idx > targ_pos {
+                let iv = Interval::new_closed_open(targ_pos, target_idx);
```

Comment:
This should probably be changed to `Interval::new` now, but will likely compile.